crml/cennzx-spot/src/types.rs

```diff
@@ -13,91 +13,144 @@
 //
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
-//!
-//! CENNZX-SPOT Types
-//!
 use cennznet_primitives::types::AssetId;
-use core::convert::TryInto;
-use uint::construct_uint;
-construct_uint! {
-	/// 128-bit unsigned integer.
-	pub struct U128(2);
+use codec::{Decode, Encode, HasCompact};
+use core::{
+	convert::{From, Into, TryFrom},
+	marker::PhantomData,
+};
+use support::decl_error;
+
+pub use primitive_types::U256 as HighPrecisionUnsigned;
+pub use u128 as LowPrecisionUnsigned;
+
+/// A trait for values which hold an implicit scale factor that needs to be taken into account in calculations
+pub trait Scaled {
+	const SCALE: LowPrecisionUnsigned;
 }
 
-construct_uint! {
-	/// 256-bit unsigned integer.
-	pub struct U256(4);
+/// Per millionth of unit price
+#[derive(Debug, Clone, Copy)]
+pub enum PerMill {}
+impl Scaled for PerMill {
+	const SCALE: LowPrecisionUnsigned = 1_000_000;
 }
 
-use codec::{Compact, CompactAs, Decode, Encode, HasCompact};
+/// Per thousandth of unit price
+#[derive(Debug, Clone, Copy)]
+pub enum PerMilli {}
+impl Scaled for PerMilli {
+	const SCALE: LowPrecisionUnsigned = 1_000;
+}
+
+/// Per hundredth of unit price
+#[derive(Debug, Clone, Copy)]
+pub enum PerCent {}
+impl Scaled for PerCent {
+	const SCALE: LowPrecisionUnsigned = 100;
+}
 
-/// FeeRate S.F precision
-const SCALE_FACTOR: u128 = 1_000_000;
+decl_error! {
+	pub enum Error {
+		Overflow,
+		DivideByZero,
+		EmptyPool,
+	}
+}
 
-/// FeeRate (based on Permill), uses a scale factor
-/// Inner type is `u128` in order to support compatibility with `generic_asset::Balance` type
-#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, Debug)]
+/// Inner type is `LowPrecisionUnsigned` in order to support compatibility with `generic_asset::Balance` type
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct FeeRate(u128);
+#[derive(Encode, Decode, Copy, Clone, Debug)]
+pub struct FeeRate<S: Scaled>(LowPrecisionUnsigned, PhantomData<S>);
 
-impl FeeRate {
-	/// Create FeeRate as a decimal where `x / 1000`
-	pub fn from_milli(x: u128) -> FeeRate {
-		FeeRate(x * SCALE_FACTOR / 1000)
+impl<S: Scaled> Default for FeeRate<S> {
+	fn default() -> Self {
+		FeeRate::<S>::from(LowPrecisionUnsigned::default())
 	}
+}
 
-	/// Create a FeeRate from a % i.e. `x / 100`
-	pub fn from_percent(x: u128) -> FeeRate {
-		FeeRate(x * SCALE_FACTOR / 100)
+impl<S: Scaled> PartialEq for FeeRate<S> {
+	fn eq(&self, other: &Self) -> bool {
+		self.0 == other.0
 	}
+}
```

Comment:
These lines can be removed because we derived `PartialEq` above

Reply:
I remember I tried deriving but couldn't pacify the compiler without explicitly implementing PartialEq

Reply:
That's because `PartialEq` was not derived for the underlying structs (PerMill, PerMilli and PerCent)

Reply:
PerMill, PerMilli, ... are empty types. There can't be any instantiation from them. That's why I don't think deriving PartialEq would do the thing we think it should. However I will test it again.

Reply:
Yeah try it. It builds and tests fine on my machine.